### PR TITLE
Qt5.15.15

### PR DESCRIPTION
**Pull request by adrianinsaval**
Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [ ] Bumped the build number (if the version is unchanged)
* [x] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [ ] Ensured the license file is being packaged.

### CODE REVIEW

**Comment by adrianinsaval** [comment]
continuation of #219

**Comment by adrianinsaval** [comment]
download of source code gets stalled :(

**Comment by adrianinsaval** [comment]
travis ci cancels out during source download, what can I do about it?

**Comment by hmaarrfk** [comment]
Hey sorry for not responding. I thought a bot had opened this due to an auto-upgrade path.

Truthfully, I am not sure this is desirable from an conda-forge ecosystem perspective.

The reason is that qtwebengine today cannot be built out easily for mac and windows.
https://github.com/conda-forge/qt-webengine-feedstock/issues/38

if you derisk qt5webengine, then we can start this upgrade.

The last migrations of qt5webengine have avoided dependencies on mac and windows, so they have been able to move forward, but having this package would create the following confusion for the solver:


1. I want the latest version of so many packages, but i can't find qtwebengine.
2. I can install old versions with qtwebengine.

We might be able to use the large machines donated by quantsight.

**Comment by adrianinsaval** [comment]
I am interested in having qt5webengine updated too. Qt has just released 5.15.13 though so we should aim for that I think. What can I do to help with that?

**Comment by hmaarrfk** [comment]
Yes. I would aim for lates version today. 

We can download the manual artifacts from azure here. And upload to our own private channels while we derisk the build system. 

The general problem with our CIs is that an error will occur at hour 16/17 and we simply can’t detect it unless we get a successful build.

**Comment by adrianinsaval** [comment]
> We can download the manual artifacts from azure here. And upload to our own private channels while we derisk the build system. 

Sorry I don't quite get what you mean here. Could you be a little more specific about what's the process here?

I have available an x86_64 Linux/windows machine and arm64 macOS machine where I can try to build locally, I can use docker on the mac to build linux_aarch64 too I think. I don't have native ppc64 but maybe it can be emulated or cross compiled?

**Comment by hmaarrfk** [comment]
The parameter you want to specify is `store_build_artifacts` 

https://github.com/conda-forge/conda-forge.github.io/blob/575274e8675a35c90d80967a92c748c8a5121551/docs/maintainer/conda_forge_yml.md?plain=1#L90

Then you can:

1. Download the artifact for linux, windows, osx
2. Upload them to your personal channel.
3. For qtwebengine
4. Add your channel to the CIs during initial development with `channel_sources`: https://github.com/conda-forge/conda-forge.github.io/blob/575274e8675a35c90d80967a92c748c8a5121551/docs/maintainer/conda_forge_yml.md?plain=1#L250
5. Get the builds to work.


It will take a while to get step 5 to work, but with your perseverance, others will come in and help!

**Comment by hmaarrfk** [comment]
For emulation, you need to install `qemu-user-static` on your linux box, then you can build whatever linux platform

It takes approximately 6x longer to emulate. So not recommended until the very end.

**Comment by adrianinsaval** [comment]
@conda-forge-admin, please rerender

**Comment by adrianinsaval** [comment]
> The parameter you want to specify is `store_build_artifacts`

it doesn't look like that's possible for travis though, guess I'll need to build qt-main aarch64 locally?

**Comment by hmaarrfk** [comment]
> it doesn't look like that's possible for travis though, guess I'll need to build qt-main aarch64 locally?

Yeah, but i have hope for this one. I can build it too and upload it. 

oddly, x86 targets older Centos by default, so you will have more challenges there.

**Comment by hmaarrfk** [comment]
Out of curiosity, what is stopping you from moving to qt6 today?

**Comment by adrianinsaval** [comment]
> Out of curiosity, what is stopping you from moving to qt6 today?

My goal in the end is to build and ship freecad/ondsel-es to users and it doesn't seem to be ready for qt6 yet. I compiled on linux with qt6 and the experience was still very buggy. Hopefully this will get fixed before end of year but for now I think it's safer to remain in qt5. vtk has also not migrated yet to qt6 and of course I also need webengine too although we plan on eventually dropping this dependency as it's very problematic.

**Comment by adrianinsaval** [comment]
so where do I download the artifacts?

**Comment by hmaarrfk** [comment]
Click on the CI results, and then "3 published" it might get to 4.
image

**Comment by adrianinsaval** [comment]
thanks, uploaded to https://anaconda.org/freecad/qt-main/files will begin trying qtwebengine maybe tomorrow. I built qt-main linux_aarch64 locally on my mac

**Comment by adrianinsaval** [comment]
I get this error when trying to build locally:
```
WARNING: failed to get install actions, retrying.  exception was: Unsatisfiable dependencies for platform osx-arm64: {MatchSpec("qt-main==5.15.13=h7e66d85_0")}
Encountered problems while solving:
  - package qt-main-5.15.13-h7e66d85_0 is excluded by strict repo priority
```

what could be going wrong?

**Comment by hmaarrfk** [comment]
are you using the `build-locally.py` script? 

you need to add your channel as a source and rerender.

If you open a PR on qt-webengine i can also review there as well.

File: recipe/meta.yaml [hunk 1]
```diff
@@ -117,7 +117,7 @@ requirements:
     - nspr                               # [unix]
     - nss                                # [unix]
     - libsqlite
-    - mysql-devel                        # [not win]
+    - mysql-devel ==8.0                  # [not win]
```
Comment (hmaarrfk) [inline recipe/meta.yaml:120]: why is this pinned? can we unpin?

Comment (hmaarrfk) [reply recipe/meta.yaml:120]: it seems maybe there wasn't a global migrator.

this definitely needs one it seems: 

Trying to add one here
https://github.com/conda-forge/conda-forge-pinning-feedstock/pull/5637

Comment (hmaarrfk) [reply recipe/meta.yaml:120]: hmm, not fun:

https://github.com/qt/qtbase/commit/7d99fa8e5bf036f97d8ef41780f6e401b32f47ef#diff-1d355c25c0b0eddec2be48253407780c4dc510d986739aec61e1ec892ccaf86e

seems like they changed the API: https://dev.mysql.com/doc/relnotes/mysql/8.3/en/news-8-3-0.html

Comment (adrianinsaval) [reply recipe/meta.yaml:120]: I tried backporting a commit that claimed to fix the build for mysql 8.3 but it didn't help. Figure it's better to update qt without updating mysql than not updating either.

**Comment by adrianinsaval** [comment]
webengine attempt: https://github.com/conda-forge/qt-webengine-feedstock/pull/48

File: recipe/meta.yaml (hunk 1 above)
Comment (hmaarrfk) [reply recipe/meta.yaml:120]: Do you mind giving https://github.com/conda-forge/qt-main-feedstock/pull/244 a quick review for the patch. I think I got it right. But as somebody who has already looked at the peticular code path your eyes are very valuable

Comment (adrianinsaval) [reply recipe/meta.yaml:120]: Sorry I don't really understand the changes and wouldn't know how to test it. That's why I gave up trying to patch it and resorted to pinning it.

Comment (hmaarrfk) [reply recipe/meta.yaml:120]: understood.

**Comment by hmaarrfk** [comment]
Typically to merge from main, i check out main
```
git checkout main
git pull upstream main
git checkout qt5.12
git rebase -i main
```
and delete all the rerender commits.

That gets rid of 90% of the merge conflicts.

**Comment by adrianinsaval** [comment]
this fell off my radar for a while but I'm back at it now with 5.15.15

**Comment by adrianinsaval** [comment]
packages available at https://anaconda.org/freecad/qt-main/files logs here: 
[qt-main-5.15.15-logs.tar.gz](https://github.com/user-attachments/files/17493685/qt-main-5.15.15-logs.tar.gz)

**Comment by adrianinsaval** [comment]
new webengine PR: https://github.com/conda-forge/qt-webengine-feedstock/pull/69

**Comment by hmaarrfk** [comment]
Thanks lets get a "stronger pin" going for qt5.15 so we can have time to build out qt-webengine.

**Comment by hmaarrfk** [comment]
I added it to the ABI migration branches.

@conda-forge-admin please rerender

**Comment by hmaarrfk** [comment]
I guess the new patch means new builds for linux.

**Comment by adrianinsaval** [comment]
yes, I don't have ppc64le hardware so that one takes long. I don't know if it's relevant there though given that the bug is for touch devices. I'm building aarch64 locally now.

**Comment by hmaarrfk** [comment]
if you have a linux server available, just start the job, then by the time you think about a long term solution, it will be done lol.

**Comment by adrianinsaval** [comment]
The patch should be ABI compatible so I don't think it's necessary to wait for ppc64le, I'll let it run tonight before sleep probably

**Comment by hmaarrfk** [comment]
> The patch should be ABI compatible so I don't think it's necessary to wait for ppc64le

no sorry, unless the requester of the PPC64le jobs steps forward, I'm not willing to drop it.

If you want to upload your own jobs to your own channel, feel free to do so. But breaking the compatibility matrix of conda-forge is not an option without a dedicated PR and some time for review from those interested.

Thank you for helping on the windows and OSX (I can't do those). I can and will start a PPC64le job.

Together we can get the full compatibility working.

I know you can likely "mix and match" qt versions with qt-webengine, but I just don't want to.

**Comment by hmaarrfk** [comment]
PPC64le should be done within 24 hours or so.

**Comment by adrianinsaval** [comment]
note that I have ppc64le built without the patch

**Comment by hmaarrfk** [comment]
> note that I have ppc64le built without the patch,

i understand, but if you are calling both the same "build number" it gets VERY confusing.

I can help on PPC64le front, and willing to do so.

> and webengine is not supported for ppc64le

Yeah that is ok. happy to work with people when they want to request it.


For example to aid during this transition to qt5.15.15 i've also added a dedicated branch for the previous version so the bot can keep it updated in case we get any delays.

**Comment by adrianinsaval** [comment]
no problem, new aarch64 build is up https://anaconda.org/freecad/qt-main/5.15.15/download/linux-aarch64/qt-main-5.15.15-h3c8598d_0.conda and here is the log: 
[linux_aarch64_.tar.gz](https://github.com/user-attachments/files/17527742/linux_aarch64_.tar.gz)

there's probably no point in copying my builds for the other platforms since they build successfully in azure

**Comment by adrianinsaval** [comment]
do we need to rebuild webengine too?

> > and webengine is not supported for ppc64le
> 
> Yeah that is ok. happy to work with people when they want to request it.

something like that would likely require massive patching, I think the problem there is that chromium itself doesn't support ppc64le, because webengine is not being built for ppac64le I think pyside is not being built for ppc64le either but I wonder if it's possible to build pyside without webengine support, I would expect yes. That can come later if there is interest

**Comment by hmaarrfk** [comment]
>  I think pyside is not being built for ppc64le

I'm not sure where the interest for qt-main came from for ppc64le. But pyside should be buildable without qt-webengine.


I merged to allow you to start building qt-webengine in a few hours.

the ppc64le build should finish on my machine "soon TM"

**Comment by hmaarrfk** [comment]
Can you please add a label "cfep03" to 
image

so i can copy that alone and not everything else?

**Comment by hmaarrfk** [comment]
command for me to run once completed:
```
anaconda copy --from-label cfep03 --to-label main --to-owner conda-forge freecad/qt-main/5.15.15
```

**Comment by adrianinsaval** [comment]
done

**Comment by adrianinsaval** [comment]
meanwhile my ppc64le build has finished and is available in the repo, here's the build log: 
[qt-main-5.15.15-linux_ppc64le_.tar.gz](https://github.com/user-attachments/files/17531099/qt-main-5.15.15-linux_ppc64le_.tar.gz)

I added the ppc label to it if you want to pull that one too

**Comment by hmaarrfk** [comment]
Thanks I uploaded both.

I guess the "wait for emulation" strategy works ^_^

**Pull request merged**